### ISSUE TITLE
Do not accept `impl` keyword by itself as a $:ty

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust#51484]
This code surprisingly compiles:

```rust
macro_rules! check {
    ($ty:ty) => {};
}

check!(impl);
```

Mentioning `impl Trait` tracking issue: #34511

**Comment by petrochenkov** [comment]
This is kind of expected, if we see `impl` in type position it must be followed by a list of bounds and that list may be empty. This is equivalent to empty bound lists in things like `fn f<T: /* nothing */>() { ... }`.

Cases like `fn f() -> impl /* nothing */ { ... }` are later rejected by a semantic check, but that check is kind of artificial, "no constraints" is equivalent to `impl Sized` which is okay if written explicitly, so I wouldn't be surprised if this restriction is lifted later.

The `macro_rules! check` ignores the type so it never goes through semantic check, so it doesn't report an error.

**Comment by dtolnay** [comment]
Thanks for the explanation, that makes sense. `impl` in type position accepts a `+`-separated list of bounds and such lists are commonly allowed to be empty elsewhere.

I had been reading `check!(impl)` as similar to `check!(*const)` which is rejected, but `*const` is not followed by a list of types.

```
error: expected type, found `<eof>`
 --> src/main.rs:5:9
  |
5 | check!(*const);
  |         ^^^^^
```
